File: wrappers/kotlin/src/main/kotlin/dev/wispers/connect/handles/Handle.kt
package dev.wispers.connect.handles

import com.sun.jna.Pointer
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Base class for all native handle wrappers.
 *
 * Provides resource management with:
 * - [requireOpen] to ensure handle is still valid before use
 * - [consume] to take ownership (for operations that invalidate the handle)
 * - [close] to release the native resource
 *
 * Implements [AutoCloseable] for use with `use {}` blocks.
 */
abstract class Handle internal constructor(
    @Volatile protected var pointer: Pointer?
) : AutoCloseable {

    private val closed = AtomicBoolean(false)

    /**
     * Whether this handle has been closed or consumed.
     */
    val isClosed: Boolean
        get() = closed.get()

    /**
     * Get the pointer, throwing if the handle has been closed.
     *
     * @throws IllegalStateException if the handle has been closed or consumed
     */
    protected fun requireOpen(): Pointer {
        if (closed.get()) {
            throw IllegalStateException("Handle has been closed or consumed")
        }
        return pointer ?: throw IllegalStateException("Handle pointer is null")
    }

    /**
     * Consume the handle, returning the pointer and marking as closed.
     *
     * Use this for operations that take ownership of the native handle
     * (e.g., logout, which frees the handle internally).
     *
     * @return The pointer, or null if already consumed
     */
    protected fun consume(): Pointer? {
        if (closed.getAndSet(true)) {
            return null
        }
        val ptr = pointer
        pointer = null
        return ptr
    }

    /**
     * Close the handle and release native resources.
     *
     * Safe to call multiple times - subsequent calls are no-ops.
     */
    override fun close() {
        val ptr = consume() ?: return
        doClose(ptr)
    }

    /**
     * Subclass-specific cleanup. Called exactly once with a valid pointer.
     */
    protected abstract fun doClose(pointer: Pointer)
}

File: wrappers/python/wispers_connect/_handle.py
"""Base handle class for opaque C pointers."""

from __future__ import annotations

import threading
from typing import Any


class Handle:
    """Base class for opaque C handle wrappers.

    Mirrors Go handle.go / Kotlin Handle.kt: tracks whether the handle has
    been closed/consumed, preventing use-after-free.
    """

    __slots__ = ("_ptr", "_closed", "_lock")

    def __init__(self, ptr: Any) -> None:
        self._ptr = ptr
        self._closed = False
        self._lock = threading.Lock()

    def _require_open(self) -> Any:
        """Return the raw pointer, raising if closed."""
        with self._lock:
            if self._closed:
                raise RuntimeError("wispers: use of closed handle")
            return self._ptr

    def _consume(self) -> Any:
        """Return the raw pointer and mark closed (for ownership-consuming calls)."""
        with self._lock:
            if self._closed:
                raise RuntimeError("wispers: use of closed handle")
            self._closed = True
            return self._ptr

    def _do_close(self, ptr: Any) -> None:
        """Subclass hook: free the C handle. Called at most once."""
        raise NotImplementedError

    def close(self) -> None:
        """Close the handle, freeing the C resource. Idempotent."""
        with self._lock:
            if self._closed:
                return
            self._closed = True
            ptr = self._ptr
        self._do_close(ptr)

    def __enter__(self) -> Handle:
        return self

    def __exit__(self, *_: object) -> None:
        self.close()

    def __del__(self) -> None:
        try:
            if not self._closed:
                self.close()
        except Exception:
            pass
